wicketd/src/lib.rs

```diff
@@ -60,22 +59,50 @@ pub async fn run_server(config: Config, args: Args) -> Result<(), String> {
         ..Default::default()
     };
 
-    let artifact_store = ArtifactStore::new(&log);
-
     let mgs_manager = MgsManager::new(&log, config.mgs_addr);
     let mgs_handle = mgs_manager.get_handle();
     tokio::spawn(async move {
         mgs_manager.run().await;
     });
 
-    let server = dropshot::HttpServerStarter::new(
+    let wicketd_server_fut = dropshot::HttpServerStarter::new(
         &dropshot_config,
         http_entrypoints::api(),
-        ServerContext { artifact_store, mgs_handle },
+        ServerContext { mgs_handle },
         &log.new(o!("component" => "dropshot (wicketd)")),
     )
     .map_err(|err| format!("initializing http server: {}", err))?
     .start();
+    tokio::pin!(wicketd_server_fut);
+
+    let artifact_server_fut = installinator_artifactd::ArtifactServer::new(
+        args.artifact_address,
+        &log,
+    )
+    .start();
+    tokio::pin!(artifact_server_fut);
+
+    installinator_artifactd::ArtifactServer::new(args.artifact_address, &log)
+        .start()
+        .await
+        .map_err(|err| format!("initializing artifact server: {:?}", err))?;
+
+    let mut wicketd_server_done = false;
+    let mut artifact_server_done = false;
 
-    server.await
+    loop {
+        tokio::select! {
+            res = &mut wicketd_server_fut, if !wicketd_server_done => {
+                wicketd_server_done = true;
+                res.map_err(|err| format!("running wicketd server: {}", err))?;
+            }
+            res = &mut artifact_server_fut, if !artifact_server_done => {
+                artifact_server_done = true;
+                res.map_err(|err| format!("running artifact server: {:?}", err))?;
+            }
```

Comment:
Q: is this the right behavior? Or should the wicketd server take priority and if it closes, the entire loop exits?

Reply:
We don't have any reason for either of these servers to "stop running", while expecting that the other needs to continue operating, right?

If that's the case, I *think* it would be fine to bail out on error if either of them fail, since that's sorta a case of either:
1. We're terminating, or
2. Something has gone horribly wrong

Reply:
Hmm, so do we expect wicketd to generally keep running forever? I guess upgrades and such would be handled by just terminating the process?

Reply:
> We don't have any reason for either of these servers to "stop running", while expecting that the other needs to continue operating, right?
> 
> If that's the case, I _think_ it would be fine to bail out on error if either of them fail, since that's sorta a case of either:
> 
> 1. We're terminating, or
> 2. Something has gone horribly wrong

I think this sounds reasonable.



> Hmm, so do we expect wicketd to generally keep running forever? I guess upgrades and such would be handled by just terminating the process?

Yes, wicketd should run indefinitely, as it's also the emergency escape hatch if the control-plane stops working. The only time it should stop is during upgrades, power outages, or the aformentioned horrible bugs.

Reply:
Thanks, I've made it so that if either server exits, the process exits.

Reply:
Note that there's a small behavior change I've introduced here:

Previously, if the wicketd server exited without an error (`Ok(())`), we'd return `Ok(())` and exit with code 0.

With this change that's no longer the case -- the server always exits with a non-zero code.